src/devices/machine/k056230.cpp

```diff
@@ -119,6 +152,316 @@ void k056230_device::ram_w(offs_t offset, u32 data, u32 mem_mask)
+		m_sock_tx.close(err);
+	m_tx_timeout.cancel();
+}
+
+void k056230_device::comm_tick()
+{
+	check_sockets();
+
+	if (m_linkenable == 0x01)
+	{
+		// if both sockets are there check ring
+		if (m_rx_state == 2 && m_tx_state == 2)
+		{
+			unsigned data_size = 0x201;
+
+			if (m_txmode == 0x00)
+			{
+				// try to read one message
+				unsigned recv = read_frame(data_size);
+				while (recv > 0)
+				{
+					// check if valid id
+					u8 idx = m_buffer0[0];
+					if (idx <= 0x0e)
+					{
+						if (idx != m_linkid)
+						{
+							// save message to ram
+							unsigned frame_start = (idx & 0x07) * 0x0100;
+							unsigned frame_size = (idx & 0x08) ? 0x0200 : 0x0100;
+							unsigned frame_offset = 0;
+							u32 frame_data = 0;
+
+							for (unsigned j = 0x00; j < frame_size; j += 4)
+							{
+								frame_offset = (frame_start + j) / 4;
+								frame_data = m_buffer0[4 + j] |  m_buffer0[3 + j] << 8 |  m_buffer0[2 + j] << 16 |  m_buffer0[1 + j] << 24;
+								ram_w(frame_offset, frame_data, 0xffffffff);
+							}
+
+							// forward message to other nodes
+							send_frame(data_size);
+						}
+					}
+
+					// try to read another message
+					recv = read_frame(data_size);
+				}
+			}
+			else if (m_txmode == 0x01)
+			{
+				// send local data to network (once)
+				unsigned frame_start = (m_linkid & 0x07) * 0x100;
+				unsigned frame_size = (m_linkid & 0x08) ? 0x200 : 0x100;
+
+				unsigned frame_offset = 0;
+				u32 frame_data = 0;
+
+				m_buffer0[0] = m_linkid;
+				for (unsigned i = 0x00; i < frame_size; i += 4)
+				{
+					frame_offset = (frame_start + i) / 4;
+					frame_data = ram_r(frame_offset, 0xffffffff);
+					m_buffer0[4 + i] = frame_data & 0xff;
+					m_buffer0[3 + i] = (frame_data >> 8) & 0xff;
+					m_buffer0[2 + i] = (frame_data >> 16) & 0xff;
+					m_buffer0[1 + i] = (frame_data >> 24) & 0xff;
+				}
+
+				send_frame(data_size);
+				m_txmode = 0x02;
+			}
+		}
+	}
+}
+
+unsigned k056230_device::read_frame(unsigned data_size)
+{
+	if (m_rx_state != 2)
+		return 0;
+
+	std::error_code err;
+	std::size_t bytes_read = m_sock_rx.receive(asio::buffer(&m_buffer0[0], data_size), asio::socket_base::message_peek, err);
+	if (err == asio::error::would_block)
+		return 0;
+
+	if (bytes_read == data_size)
+		bytes_read = m_sock_rx.receive(asio::buffer(&m_buffer0[0], data_size), 0, err);
+
+	if (err)
+	{
+		osd_printf_verbose("k056230: RX error receiving - %d %s\n", err.value(), err.message());
+		m_sock_rx.close(err);
+		m_rx_state = 0;
+		return 0;
+	}
+
+	if (bytes_read == data_size)
+		return bytes_read;
+	return 0;
+}
+
+void k056230_device::send_frame(unsigned data_size)
+{
+	if (m_tx_state != 2)
+		return;
+
+	std::error_code err;
+	std::size_t bytes_sent = m_sock_tx.send(asio::buffer(&m_buffer0[0], data_size), 0, err);
+	if (err || bytes_sent != data_size)
+	{
+		osd_printf_verbose("k056230: TX error sending - %d %s\n", err.value(), err.message());
+		m_sock_tx.close(err);
+		m_tx_state = 0;
+	}
+}
```

Comment:
Using `send` here is problematic for the same reasons that MAME’s sockets are problematic.  If you use `send` with a non-blocking socket, you still need to keep a local buffer and retry writes when the socket has buffer space available, which is the part that ASIO is supposed to help you deal with.

You want to use `async_transmit_some` to get as much on the wire as possible, then when you get the callback, call it again as long as you still have buffered data to transmit.  All you should be doing here is buffering the data and starting an `async_transmit_some` if you don’t already have one in flight.